enhancements/machine-api/maintenance-lock.md

```diff
@@ -0,0 +1,232 @@
+  - "@beekhof"
+reviewers:
+  - @bison
+  - @enxebre
+  - @michaelgugino
+  - @derekwaynecarr
+approvers:
+  - TBD
+creation-date: 2019-12-06
+last-updated: 2019-12-09
+status: provisional
+see-also:
+  - https://github.com/kubernetes/enhancements/pull/1080
+replaces:
+  - https://github.com/kubevirt/node-maintenance-operator
+superseded-by:
+  - NA
+---
+
+# Maintenance Lock
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+1. Should UIs/consoles set the annotation directly, or signal for an operator to do so by creating a CR?
+
+## Summary
+
+There is a need for a common place to request and watch for nodes to be put
+into maintenance.  Whether maintenance mode is required to perform triage or
+hardware replacement, a standardized way to represent it allows system and
+layered components to delay activities that would inhibit those activities and
+for UIs/consoles to communicate status to admins.
+
+On baremetal, there is a need to perform additional tasks, such as draining the
+node, when maintenance is requested.  However web UIs do not have access to the
+`oc` command, preventing them from invoking a Go based drain library.  This
+forces admins to use the `oc` command to trigger a client-side cordon and/or
+drain of nodes, as well as to construct additional commands to determine
+progress, wait for completion, triage failures, and unwind the process once the
+maintenance activity is completed.
+
+## Motivation
+
+Badly timed updates can make triage an even longer process, but poorly timed
+power events could corrupt the machine and represent a health and safety risk.
+
+Without first-class support from the server, the manual steps needed to perform
+equivalent functionality may lack consistency between admins, present
+opportunities for mistakes, and are not reflected in the UI/consoles for other
+parties to base informed decisions.
+
+All these aspects are especially important in Telco environments where there
+are multiple levels of admins, across different physical locations, managing
+thousands of clusters.
+
+In such an environment, the admin needs to find identify the cluster a
+problematic node belongs to as well as locate and use the corresponding
+`kube.config` and/or pass `oc` the correct cluster name for every call.  Not
+impossible but certainly error prone.
+
+### Goals
+
+- A mechanism for signaling that a node is in a maintenance state
+
+- A metal3 specific operator that performs relevant actions (eg. cordon +
+  drain) when maintenance is requested
+
+- Changes to existing components to disable or delay functions that would
+  affect the node, such as:
+  - Delay upgrades
+  - Disable health checks
```

Comment:
what about autoscaling? machineSet/machineDeployment scaling-in/deletion operation? any other upper level controller that manipulate machines/nodes?
Is the idea here for every controller to decide if they should honour `lifecycle.openshift.io/maintenance`? If so can we try to elaborate and to be granular on what we are trying to protect against here? is it just rebooting, is it also node deletion, machine deletion?

Reply:
If I could wave a magic wand, I would prevent any event that came from the system and resulted in machines being powered off or deleted.

That being said, trying to determine whether the event came from an admin or the system is a non-trivial problem, and the feature can still be considered useful if only a subset of controllers respect the annotation.

However based on Michael's conversation with Clayton overnight, I will need to rewrite this enhancement based on leases as the signaling mechanism ( https://github.com/kubernetes/enhancements/issues/1403 ) and will make sure to include how scaling/deletion is affected

Reply:
>If I could wave a magic wand, I would prevent any event that came from the system and resulted in machines being powered off or deleted.

As a thought that could be achieved for deletion (or any supported semantic by the core API, may be in future poweroff) by moving the responsibility to the core controller with a finaliser (i.e machine controller) rather than put it into the consumer, i.e the machine controller on a deletion event wouldn't finalise until the lease for the backed node is gone. How ever not sure that's a good idea.

> and the feature can still be considered useful if only a subset of controllers respect the annotation.

Absolutely I'm suggesting we try to put a definition into words so when I'm developing my controller I can see if it fits.

Reply:
I believe this is addressed in the latest version